src/libraries/Common/src/System/Net/ArrayBuffer.cs

```diff
@@ -134,72 +166,24 @@ public void EnsureAvailableSpace(int byteCount)
             _activeStart = 0;
 
             _bytes = newBytes;
-            if (_usePool)
-            {
-                ArrayPool<byte>.Shared.Return(oldBytes);
-            }
+            ReturnBufferIfPooled(oldBytes);
 
             Debug.Assert(byteCount <= AvailableLength);
         }
 
-        // Ensure at least [byteCount] bytes to write to, up to the specified limit
-        public void TryEnsureAvailableSpaceUpToLimit(int byteCount, int limit)
+        public void Grow()
         {
-            if (byteCount <= AvailableLength)
-            {
-                return;
-            }
-
-            int totalFree = _activeStart + AvailableLength;
-            if (byteCount <= totalFree)
-            {
-                // We can free up enough space by just shifting the bytes down, so do so.
-                Buffer.BlockCopy(_bytes, _activeStart, _bytes, 0, ActiveLength);
-                _availableStart = ActiveLength;
-                _activeStart = 0;
-                Debug.Assert(byteCount <= AvailableLength);
-                return;
-            }
-
-            if (_bytes.Length >= limit)
-            {
-                // Already at limit, can't grow further.
-                return;
-            }
-
-            // Double the size of the buffer until we have enough space, or we hit the limit
-            int desiredSize = Math.Min(ActiveLength + byteCount, limit);
-            int newSize = _bytes.Length;
-            do
-            {
-                newSize = Math.Min(newSize * 2, limit);
-            } while (newSize < desiredSize);
-
-            byte[] newBytes = _usePool ?
-                ArrayPool<byte>.Shared.Rent(newSize) :
-                new byte[newSize];
-            byte[] oldBytes = _bytes;
-
-            if (ActiveLength != 0)
-            {
-                Buffer.BlockCopy(oldBytes, _activeStart, newBytes, 0, ActiveLength);
-            }
-
-            _availableStart = ActiveLength;
-            _activeStart = 0;
-
-            _bytes = newBytes;
-            if (_usePool)
-            {
-                ArrayPool<byte>.Shared.Return(oldBytes);
-            }
-
-            Debug.Assert(byteCount <= AvailableLength || desiredSize == limit);
+            EnsureAvailableSpaceCore(AvailableLength + 1);
         }
 
-        public void Grow()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ReturnBufferIfPooled(byte[] buffer)
         {
-            EnsureAvailableSpace(AvailableLength + 1);
+            // The buffer may be Array.Empty<byte>()
```

Comment:
ArrayPool is fine with Array.Empty being returned; it just gets ignored.

Reply:
I'll remove the extra check, given that we won't be calling `ClearAndReturnBuffer` on already-empty buffers often.